### PR TITLE
Add IAM

### PR DESCRIPTION
**Pull request by illi-j**
This PR add IAM to egoscale.

### CODE REVIEW

File: iam_apikey.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+	Key         string     `json:"key"`
+	Secret      string     `json:"secret,omitempty"`
+	Operations  []string   `json:"operations,omitempty"`
+	Type        APIKeyType `json:"type"`
+}
+
+// CreateAPIKey represents the apikey creation
+type CreateAPIKey struct {
+	Description string `json:"description"`
+	Operations  string `json:"operations,omitempty"`
```
Comment (pierre-emmanuelJ) [inline iam_apikey.go]: I think can be nice to put a slice for this field

File: iam_apikey.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
+package egoscale
+
+// APIKeyType holds the type of the apikey
+type APIKeyType string
+
+const (
+	// APIKeyUnrestricted is unrestricted
+	APIKeyUnrestricted APIKeyType = "restricted"
+	// APIKeyRestricted is restricted
+	APIKeyRestricted APIKeyType = "unrestricted"
```
Comment (falzm) [inline iam_apikey.go]: I suggest naming the constants `APIKeyTypeRestricted`/`APIKeyTypeUnrestricted` so one can know at a glance what they refer to.

File: iam_apikey.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+)
+
+// APIKey represents an apikey
```
Comment (falzm) [inline iam_apikey.go]:
```suggestion
// APIKey represents an API key
```

File: iam_apikey.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
+package egoscale
+
+// APIKeyType holds the type of the apikey
```
Comment (falzm) [inline iam_apikey.go]:
```suggestion
// APIKeyType holds the type of the API key
```

File: iam_apikey.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+// APIKey represents an apikey
+type APIKey struct {
+	Description string     `json:"description"`
+	Key         string     `json:"key"`
+	Secret      string     `json:"secret,omitempty"`
+	Operations  []string   `json:"operations,omitempty"`
+	Type        APIKeyType `json:"type"`
+}
+
+// CreateAPIKey represents the apikey creation
```
Comment (falzm) [inline iam_apikey.go]:
```suggestion
// CreateAPIKey represents the API key creation
```

File: iam_apikey.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+	Operations  string `json:"operations,omitempty"`
+	_           bool   `name:"createApiKey" description:"Create an apikey."`
+}
+
+// Response returns the struct to unmarshal
+func (CreateAPIKey) Response() interface{} {
+	return new(APIKey)
+}
+
+// ListAPIKeys represents a search for an apikeys
```
Comment (falzm) [inline iam_apikey.go]:
```suggestion
// ListAPIKeys represents a search for an API key
```

File: iam_apikey.go [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+type ListAPIKeys struct {
+	_ bool `name:"listApiKeys" description:"List apikeys."`
+}
+
+// ListAPIKeysResponse represents a list of apikeys
```
Comment (falzm) [inline iam_apikey.go]:
```suggestion
// ListAPIKeysResponse represents a list of API keys
```

File: iam_apikey.go [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+	Count   int      `json:"count"`
+	APIKeys []APIKey `json:"apikeys"`
+}
+
+// Response returns the struct to unmarshal
+func (ListAPIKeys) Response() interface{} {
+	return new(ListAPIKeysResponse)
+}
+
+// RevokeAPIKey represents a revoke of apikey
```
Comment (falzm) [inline iam_apikey.go]:
```suggestion
// RevokeAPIKey represents a revocation of an API key
```

File: iam_apikey.go [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+type RevokeAPIKey struct {
+	Key string `json:"key"`
+	_   bool   `name:"revokeApiKey" description:"Revoke an apikey."`
+}
+
+// RevokeAPIKeyResponse represents a revoke of apikey
```
Comment (falzm) [inline iam_apikey.go]:
```suggestion
// RevokeAPIKeyResponse represents the response to an API key revocation
```

File: iam_apikey.go (hunk 1 above)
Comment (falzm) [reply iam_apikey.go]: I agree.

**Comment by falzm** [comment]
If you PR is ready to be reviewed you can remove the WIP label ;)

File: iam_apikey.go (hunk 2 above)
Comment (greut) [reply iam_apikey.go]: Also, they are obviously swapped.

File: iam_apikey.go (hunk 1 above)
Comment (illi-j) [reply iam_apikey.go]: not possible for the API at the moment.
(likely to change in the near future)

File: iam_apikey.go [hunk 10]
```diff
@@ -47,10 +47,25 @@ func (ListAPIKeys) Response() interface{} {
 	return new(ListAPIKeysResponse)
 }
 
+// ListAPIKeyOperations represents a search for operations for the current API key
```
Comment (falzm) [inline iam_apikey.go]: Is it really necessary to implement this call? Doesn't the `getAPIKey` also return the key's supported operations?

Comment (alexanderjamesking) [reply iam_apikey.go]: There is a difference.

ListApiKeyOperations returns all of the things a key can do (even for a key without restrictions - you get the full list), whereas GetApiKey returns the details of the policy for the key - so for an unrestricted key no operations would be returned.

ListApiKeyOperations is primarily for the UI so we can display a list of checkboxes when creating a new key, it's debatable how useful this is from the CLI as the operations do not always align with the commands exposed by the CLI. If we were to do some autocomplete on the operations from the CLI you could use it.

**Pull request merged**